toothpick-runtime/src/main/java/toothpick/UnScopedProviderImpl.java

```diff
@@ -44,60 +46,65 @@ public UnScopedProviderImpl(Class<?> factoryKeyClass, boolean isProviderFactoryC
 
   //we lock on the unbound provider itself to prevent concurrent usage
   //of the unbound provider (
-  public synchronized T get(Scope scope) {
-    if (instance != null) {
-      return instance;
-    }
-
-    if (providerInstance != null) {
-      if (isLazy) {
-        instance = providerInstance.get();
-        //gc
-        providerInstance = null;
+  public T get(Scope scope) {
+    lockGet.lock();
```

Comment:
http://blog.takipi.com/java-8-stampedlocks-vs-readwritelocks-and-synchronized/

this one says synchronized is better for a low number of threads, which will be the case in Android. LOL. No one agrees :) 
I would keep the lock until someone else comes with a proper synchronized approach and benchmarks to say it's better... Apparently we got an improvement in speed using locks and the new lock free DS is actually well synchronized I believe. So it cannot really be compared to what we used to have which was not thread safe, or not totally.